### DISCUSSION TITLE
How to use vercel nameservers in cloudflare?

### DISCUSSION THREAD

**Discussion by waptik** [vercel/vercel#7739] [Help] [answered]
Hi. I have a domain name that I bought from a domain name provider and I setup Cloudflare as my DNS provider. I added my domain name to my vercel project as a wildcard domain name. I'm basing my project on https://vercel.com/guides/nextjs-multi-tenant-application#5.-deploy-to-vercel
 How do I insert the following nameservers in Cloudflare? Or do I have to forgo cloudflare and use prokbun as my DNS provider?
```
ns1.vercel-dns.com
ns2.vercel-dns.com
```

**Comment by EryouHao** [comment]
Any progress on this issue? @waptik

**Reply by waptik** [reply]
Nope. The only solution is to have another domain(eg. `notaku.site`) exclusively have vercel nameservers so that you can use wildcard subdomain(`blog.notaku.site`) and your primary domain(eg. `notaku.website`) on cf for frontpage and site management

**Comment by amyegan** [comment]
Edit: See latest answer here: https://github.com/vercel/vercel/discussions/7739#discussioncomment-6632066

https://vercel.com/docs/integrations/cloudflare

**Reply by mbalesni** [reply]
Hey @amyegan! Do you have any tips for what to do if one followed the support article on [how to use Cloudflare with Vercel](https://vercel.com/support/articles/using-cloudflare-with-vercel) with the [proxy](https://vercel.com/support/articles/using-cloudflare-with-vercel#with-proxy) option, the domain works fine, but Vercel stills shows the configuration is invalid (and regularly sends the warning emails)? 

If I use DNS-only and Flexible SSL, Vercel says the configuration is good. In both cases, the website works fine.

**Reply by amyegan** [reply]
Hi @nikebless! Are you using a CNAME record with the target set to `cname.vercel-dns.com` or A record with IPv4 address set to `76.76.21.21`? You may have already tried this, but it may help to remove the domain from your project and add it again. Good old "turn it off and on again" approach 😉 

If that doesn't work, I'll need to know the domain you're trying to use so I can look into this issue further. If you need to keep that information secret, you can [submit a case](https://vercel.com/help#issues) and the Customer Success team can help you privately.

**Reply by mbalesni** [reply]
I have tried the CNAME and A record and now tried the good old off-and-on approach, still getting the "Invalid Configuration" note. I would like to talk to support, but will they talk to me about this while I'm on the Hobby plan? I got the impression from the [Support Terms](https://vercel.com/legal/support-terms) that they won't :)

**Reply by benknab** [reply]
I'm getting the same error. Record works fine though, my app is loading as expected but Vercel is complaining about the missing CNAME record

**Reply by amyegan** [reply]
Can you share more details about what you're trying to add and what errors you're getting? If you are using subdomains, you'll need to add records for each of them as well as the apex domain. A walk-through, in case you didn't already see it, can be found at [vercel.com/docs/concepts/projects/domains/add-a-domain](https://vercel.com/docs/concepts/projects/domains/add-a-domain). Be sure to also set the SSL/TLS mode to "Full" to encrypt all traffic if [using Cloudflare proxy](https://vercel.com/guides/using-cloudflare-with-vercel#with-proxy)

**Reply by benknab** [reply]
yeah, I followed the docs, the subdomain has the CNAME record, Cloudflare has SSL on Full, and everything works correctly, but Vercel is sending me false positive error emails about my subdomains having an invalid configuration and in the Domains tab it also shows up with the invalid configuration error

but let me repeat that I do not experience any errors, it's just the message showing up

**Reply by amyegan** [reply]
Sorry to hear you're still getting incorrect error messages @bkiac. Can you [start a new discussion](https://github.com/vercel/community/discussions/new) or [submit a case](https://vercel.com/help#issues) to the Customer Success Team so we can work together to solve this without pinging everyone else in this discussion?

**Reply by benknab** [reply]
thank you @amyegan, I submitted a ticket

**Reply by codewithmilo** [reply]
Just experienced this same issue: CNAME record set up, site works properly, but getting invalid configuration issues from namespaces that I can't change. Was there any solution discovered?

**Comment by leerob** [comment]
P.S. we have new docs for nameservers! https://vercel.com/docs/concepts/projects/domains/managing-nameservers

**Reply by waptik** [reply]
> P.S. we have new docs for nameservers! [vercel.com/docs/concepts/projects/domains/managing-nameservers](https://vercel.com/docs/concepts/projects/domains/managing-nameservers)

The new documentation is related to the scope of this discussion. If i'm wrong and it is related then there's not enough information on how to proceed.

**Comment by akhil-naidu** [comment]
@amyegan, when using Cloudflare as my domain registrar, is it impossible to use Vercel wildcard domains?

I cannot edit my nameservers in Cloudflare but can add NS records in Cloudflare, so is it possible to add wild card support using these NS records?

**Reply by amyegan** [reply]
Good question, @akhil-naidu. [Wildcard domains](https://vercel.com/blog/wildcard-domains) must use the Vercel nameservers method of adding custom domains. This is because [Vercel needs to be able to set DNS records](https://vercel.com/guides/why-use-domain-nameservers-method-wildcard-domains#why-is-the-nameservers-method-required-here) to generate wildcard certificates.

**Reply by raRaRa** [reply]
Any reason why this is a requirement? AWs + Cloudflare makes this work just fine, e.g. make Cloudflare point *.domain.com to a CloudFront distribution.

**Reply by amyegan** [reply]
It's a requirement for [SSL certification](https://vercel.com/docs/concepts/projects/domains#ssl-certification):

> Vercel uses LetsEncrypt for certificates. For all non-wildcard domains, we use the [HTTP-01 challenge method](https://letsencrypt.org/docs/challenge-types/#http-01-challenge) and providing the request can make it to Vercel, then our infrastructure will deal with it. For wildcard requests, we use the [DNS-01 challenge method](https://letsencrypt.org/docs/challenge-types/#dns-01-challenge). This is why we require nameservers to be with Vercel to use wildcard domains - if the DNS isn't with us, we can't make the DNS record to approve it.

**Reply by raRaRa** [reply]
@amyegan Thank you! Is there an alternative solution to this, because Cloudflare doesn't allow people to use a different name server for domains registered through them. One potential solution I have in mind is to use URL rewrite for the hostname on the Cloudflare side, so that {customer}.domain.com rewrites to domain.com/c/{customer} - that way I could also skip the middleware layer in NextJS. Are there any potential cons to this?

**Reply by amyegan** [reply]
You could use rewrites, or add customer subdomains as need, instead of using a wildcard. The downside is the extra work of setting up  each subdomain if you need to do it often. Not a huge downside if new subdomains don't need to be created frequently.

Alternatively, if you think wildcard domains would make things much easier for your app, you can [transfer](https://developers.cloudflare.com/registrar/account-options/transfer-out-from-cloudflare/) the domain either [to Vercel](https://vercel.com/guides/transferring-domains-to-vercel) or to another registrar. [Namecheap](https://www.namecheap.com/), [Google Domains](https://domains.google/), and [Hover](https://www.hover.com/) are popular recommendations.

**Reply by akhil-naidu** [reply]
@raRaRa If you consider using subdomains as individual applications, I see no issues. But when you use NextJS for a multi-tenant application, I prefer adding the necessary subdomains via an API call. Configuring wildcards would be better for this API call to create a subdomain and SSL.

@amyegan Though I crossed the 60 days domain limit to transfer it to another registrar, I'm a premium user of Cloudflare, which provides us with many other benefits like Web application Firewall and Page Rules. So migrating my register would be my least priority.

So, for now, I created a wildcard CNAME record pointing at `cname.vercel-dns.com`, and everything seems fine, at least for using this with [Vercel Platforms Starter Kit.](https://github.com/vercel/platforms)

**Reply by raRaRa** [reply]
I ended up URL re-writing through CloudFront.

Cloudflare -> CloudFront (_re-writes {customer}.domain.com to domain.com/c/{customer}_) -> Vercel

This way I don't need to create subdomain per customer. However I still needed NextJS middleware to URL rewrite client-side requests.

Once Cloudflare snippets is released I will switch to Cloudflare -> Vercel instead of using CloudFront in the middle.

**Reply by amyegan** [reply]
Thanks for sharing that @akhil-naidu. I hadn't had success with that in the past, but maybe it's a premium feature or something changed since I last tried it. I'm happy to hear the wildcard CNAME record is working for you! Looks like I need to try it again. 😄

And thank you @raRaRa for sharing your solution as well. I'm glad the CloudFront rewrite works for you!

**Comment by smaeda-ks** [comment]
For those who landed on this old community post, let me update the previously marked answer, as it contains some inaccurate information as of today.

First of all, you **_can_** use a wildcard domain on your Vercel project with Cloudflare name servers. Our [documentation](https://vercel.com/guides/why-use-domain-nameservers-method-wildcard-domains), as well as the UI's error message, are somewhat misleading, but you can "ignore" the error message as long as you follow the below steps.

### 1. Add your wildcard domain to your project

First, you add `*.example.com` to your project (`Project Settings` -> `Domains`). At this stage, you will see the `Invalid Configuration` error message on the UI instructing you to change its name servers to ours. **Please ignore this**, you will be fine.

### 2. Issue your wildcard TLS certificate via Vercel CLI

This is the most crucial step. Since we use Let's Encrypt certificates, there are two ways to verify your domain ownership:
https://letsencrypt.org/docs/challenge-types/

Because this is a "wildcard" certificate, you can only verify via the DNS-01 challenge mentioned in the above documentation. This is why we initially said it's necessary to change the name servers so that we can automatically handle this DNS-01 challenge on our side without having you manually create a verification TXT DNS record. But that's not a hard requirement. You can still do this DNS-01 challenge manually using the Vercel CLI:
https://vercel.com/docs/cli/certs#challenge-only

```sh
$ vercel certs issue "*.example.com" --challenge-only
Vercel CLI 28.12.6
> A certificate issuance for *.example.com has been started [2s]
  Add the following TXT records with your registrar to be able to the solve the DNS challenge:

name                   type        value
_acme-challenge        TXT         YN7LvYccJ0Xlrih9lIb--gcgH8OzlsD4POWf15XtrDc

> To issue the certificate once the records are added, run:
  `vercel certs issue *.example.com`
  Read more: https://err.sh/vercel/solve-challenges-manually

$ vercel certs issue "*.example.com"
Vercel CLI 28.12.6
> Success! Certificate entry for *.example.com created [6s]
```

^^ This is the example output. You can then create a new TXT record via the Cloudflare DNS interface with the provided value.

**[Important Note]**

Because Let's Encrypt certificates have 90 days expiration period, and you will need a new verification challenge signature for the certificate auto-renewal, you have to follow the same step and re-issue the certificate and update the TXT record with the new verification signature before the cert expires. Please see my follow-up comment below for more details.

### 3. Create a wildcard DNS entry in Cloudflare

Once step 2 is completed, you can add a new [wildcard DNS entry](https://developers.cloudflare.com/dns/manage-dns-records/reference/wildcard-dns-records/) via the Cloudflare DNS interface, and point it to our shared CNAME hostname `cname.vercel-dns.com`. So it looks something like this:

<img width="1329" alt="Screen Shot 2023-01-18 at 23 07 51" src="https://user-images.githubusercontent.com/11495867/213193341-d7c6ecf4-d5e2-44b9-8f79-892bf4126205.png">

### 4. (Optional) Disable warning notifications (UI's Inbox notification/Email)

Note that because of how we're checking the domain's name servers today, the `Invalid Configuration` error message will continue to appear even though your setup is perfectly fine.

<img width="943" alt="Screen Shot 2023-01-18 at 23 26 23" src="https://user-images.githubusercontent.com/11495867/213196903-2eaceb3f-f965-43ea-9328-32b3208330a4.png">

Until we have better UI/UX around this name server check, you can disable warning notifications via https://vercel.com/account/notifications so we won't bother you.

- `Web` -> `Domains` -> `Configuration`
- `Email` -> `Domains` -> `Configuration`

**Reply by raRaRa** [reply]
Thank you so much for this clarification, and this makes 100% sense. I ended up using CloudFront in front of Vercel to bypass this issue. Cloudflare -> CloudFront (basically rewrites the request from *.domain.com to domain.com) -> Vercel. Really glad to be able to skip CloudFront with this solution :)

**Reply by smaeda-ks** [reply]
My apologies. I spoke too soon. I also had a misunderstanding and fixed my answer above.

Basically, as of today, we can't renew wildcard certificates automatically if the domain's name servers are not pointing to Vercel. Therefore, you will need to manually re-issue your wildcard certificate every ~90 days (all Let's Encrypt certs have 90 days expiration period) using the Vercel CLI as highlighted above.

The main reason is that renewing a Let's Encrypt cert requires a new verification signature; thus, the initial TXT record won't work for the next renewal, unfortunately.

Yes, the workflow is still cumbersome and should be streamlined; we acknowledge that. I just chatted with our engineering team, and we do think there's a way to eliminate this limitation. So I'm going to raise an internal ticket and pursue that route, but there are some challenges, and it will need enough engineering time investment, which may take some time to work on. We will keep you updated on this thread.

In the meantime, I thought I'd also mention another workaround you may be able to approach. In this case, you can also upload your custom certificates:
https://vercel.com/docs/concepts/projects/domains/custom-SSL-certificate
(note that it is only available to the Enterprise plan)

Since Cloudflare allows you to issue custom certificates using their [Origin Certificate](https://developers.cloudflare.com/ssl/origin-configuration/origin-ca/) offering with a much longer expiration period (up to 15 years), you could upload it to Vercel to reduce the operational burden, so you don't have to update the certificate every XX days.

We can also discuss potentially making the custom cert feature available to other plans (Pro and Hobby) as an alternative approach.

**Reply by smaeda-ks** [reply]
Also, Cloudflare seems to support custom name servers, and it's available to `Business` and `Enterprise` plans as per:
https://developers.cloudflare.com/dns/additional-options/custom-nameservers/

**Reply by smaeda-ks** [reply, marked as answer]
For those folks landing this page today, we now have a workaround allowing you to set up and renew your wildcard certificates without switching to Vercel name servers. Please follow this guide: https://vercel.com/guides/wildcard-domain-without-vercel-nameservers

tl;dr; we started allowing the NS delegation method for the DNS-01 challenge.

https://letsencrypt.org/docs/challenge-types/#dns-01-challenge

> Since Let’s Encrypt follows the DNS standards when looking up TXT records for DNS-01 validation, you can use CNAME records or NS records to delegate answering the challenge to other DNS zones. This can be used to [delegate the _acme-challenge subdomain](https://www.eff.org/deeplinks/2018/02/technical-deep-dive-securing-automation-acme-dns-challenge-validation) to a validation-specific server or zone.

### [update on 8/30/2023]

The above workaround only partially supports setting up a wildcard cert with Cloudflare proxy mode.

tl;dr; Cloudflare also uses the same `_acme-challenge` DNS label for their cert issuance, which conflicts with this workaround, making them unable to issue certs correctly.

Therefore, we recommend not using Cloudflare proxy mode if you need to set up a wildcard cert on Vercel. Consider using DNS-only mode instead.

Otherwise, you will need to manipulate your DNS record periodically to make it work, or upload a custom cert to Cloduflare, which still requires you to update and upload the new cert manually periodically.

**Reply by sanderfish** [reply]
@smaeda-ks this looks promising! I've started setting this up. Without proxy, the wildcard subdomain works!

As soon as I turn on the proxy; domain verification in the project in Vercel fails.

This is what I've done so far

- Added wildcard domain to Cloudflare: CNAME | *.domain.com | cname.vercel-dns.com
- Added _acme-challenge NS records to Cloudflare 
- Added wildcard domain to Domains *.domain.com
- Cloudflare SSL is set to full (APEX domain uses Cloudflare proxy and runs on Vercel without issue)

I don't see the "Enable Vercel Domains" button, which I think is the issue.

<img width="1131" alt="CleanShot 2023-08-08 at 16 34 03@2x" src="https://github.com/vercel/vercel/assets/7306204/4bc3b52f-c7d9-44bf-b9c1-14720d7df2c5">

**Reply by smaeda-ks** [reply]
@sanderfish Vercel DNS is already enabled for that domain and that's why you don't see the button. Your wildcard cert seems to be available now: `$ sslscan --sni-name="*.checkoutpage.co" cname.vercel-dns.com`

**Reply by sanderfish** [reply]
@smaeda-ks thanks for the reply! I'm testing this on another domain and while it is working, with Cloudflare proxy turned on, I'm still seeing the following error in the dashboard.

<img width="942" alt="image" src="https://github.com/vercel/vercel/assets/7306204/57932c33-cf76-498d-828f-a9742a0d616e">

<img width="1249" alt="image" src="https://github.com/vercel/vercel/assets/7306204/7dc76462-2f40-4148-a5f4-511bd98313f5">

With the proxy turned on, I'm also unable to verify custom domains via the CNAME record, such as:

checkout.otherdomain.com CNAME cname.checkoutfront.com 

As soon as I turn the proxy off, the verification goes through.

**Reply by kaumac** [reply]
@sanderfish what solution did you decide on? Curious as I'm facing the same problem.

I'm subscribing to this to hopefully get updates when a solution rolls out. I'll be using the vercel nameservers for now as I don't think there's much benefit in using cloudflare without proxy.
